Add header component

alex-ju opened the pull request:
Review link: https://govuk-frontend-review.herokuapp.com/components/header

New version of the GOV.UK Header based off the work done in the GOV.UK Design System.

## Notable changes
- Macro interface that supports service name and navigation
- B.E.M.
- Updates spacing based on the existing spacing scale and values
- Fixes alignments
- Uses inline SVG for the crown (can be see in inverse, prints sharp)
- Mobile navigation items on a single line with appropriate touch area

## Testing
[browser and AT tests were done; screen captures and notes coming soon]

## Notes
Although some experimental work was done around it, this PR doesn't include the search component and the macro interface for it.

alex-ju commented on src/header/_header.scss:
```diff
@@ -0,0 +1,290 @@
+@import "../globals/helpers/typography";
+@import "../globals/settings/typography-font-stacks";
+@import "../globals/settings/typography-font";
+@import "../globals/settings/typography-responsive";
```
yep. will probably be merged after the restructure anyway. will be easier to rebase on this one.

igloosi commented:
padding top on ie8 adds a black line 
![screen shot 2018-05-15 at 10 29 16](https://user-images.githubusercontent.com/3758555/40048737-de02acd8-582a-11e8-92fd-8f9d6de8c1db.png)

igloosi commented:
firefox doesnt seem to like toggling
![screen shot 2018-05-15 at 10 35 19](https://user-images.githubusercontent.com/3758555/40049106-b12abe3e-582b-11e8-8c09-cd5bd73106ca.png)

igloosi commented:
looks good and works in other supported browsers

igloosi commented on src/header/template.njk:
```diff
@@ -0,0 +1,61 @@
+  <div class="govuk-header__container {{ params.containerClasses if params.containerClasses }}">
+
+    <div class="govuk-header__logo">
+      <a href="{{ params.homepageUrl }}" class="govuk-header__link govuk-header__link--homepage">
```
could we consider having a default of / here?

alex-ju replied:
Updated, added `/` as default value.

igloosi replied:
👍

igloosi commented on src/header/_header.scss:
```diff
@@ -0,0 +1,290 @@
+  }
+
+  .govuk-header__navigation--no-service-name {
+    padding-top: 40px; // 30px line-height of the service name + 10px for margin-bottom
```
$govuk-spacing-scale-7 ?

alex-ju replied:
Updated, used the recommended `$govuk-spacing-scale-7` for padding.

igloosi replied:
👍

36degrees commented on src/header/_header.scss:
```diff
@@ -0,0 +1,290 @@
+  .govuk-header__menu-button--open {
+    &::after {
+      @include govuk-shape-arrow($direction: up, $base: 10px);
+      content: "";
```
Shouldn't need to redeclare `content`, `display` or `margin-left`, they should be able to inherit from `govuk-header__menu-button`?

alex-ju replied:
Updated, removed attributes that could be inherited. `display` is still needed, as it overwrites the style coming from the `govuk-shape-arrow` mixin.

alex-ju commented on src/header/template.njk:
```diff
@@ -0,0 +1,61 @@
+            height="32"
+            width="36"
+          >
+            <path
```
Updated. Added the comments from the Design System.

igloosi commented:
https://github.com/alphagov/govuk-frontend/pull/695/commits/fc0b5d93f8e93a9c3add9289dacd684184c98f40 fixed the Firefox issue

alex-ju commented:
Updated, addressing all comments.

@igloosi:
- reeeeally long service name was addressed in c38e15f
- padding top on ie8 adds a black line fixed with bce4888
- firefox on windows doesn't seem to like toggling fixed with fc0b5d9

alex-ju commented:
Restructured based on changes coming from #693.

alex-ju commented on src/components/header/_header.scss:
```diff
@@ -0,0 +1,274 @@
+
+  $govuk-header-background: $govuk-black;
+  $govuk-header-border-color: $govuk-blue;
+  $govuk-header-border-width: $govuk-spacing-scale-2;
```
We could, but that will make it inconsistent across components. (i.e. variables declaration in footer)

igloosi replied:
let's keep it consistent then

36degrees replied:
[We're](https://github.com/alphagov/govuk-frontend/blob/master/src/components/breadcrumbs/_breadcrumbs.scss#L7-L20) [already](https://github.com/alphagov/govuk-frontend/blob/master/src/components/button/_button.scss#L10) [inconsistent](https://github.com/alphagov/govuk-frontend/blob/master/src/components/footer/_footer.scss#L9-L14) 😞 

¯\\_(ツ)_/¯

igloosi replied:
:shipit:

alex-ju replied:
I'd rather raise a separate PR to address this inconsistencies, but as you wish..

36degrees replied:
Totally happy for that to happen, just pointing out that the inconsistency is already there. Don't expect it to be addressed in this PR.

alex-ju replied:
Yep, you're right. It was probably introduced with the footer and I carried on with it.

hannalaakso commented on app/views/examples/header-and-footer/index.njk:
```diff
@@ -0,0 +1,48 @@
+{# {{ govukSkipLink({
+  href: '#content',
+  text: 'Skip to main content'
+}) }} #}
```
Why is the code commented out? Is it opt in by uncommenting it? Could the code mention this?

alex-ju replied:
Wasn't sure if the example should contain the skip-link or not. I added it back now. I think is useful when testing. Would be good to have the cookie-banner added to when it will be in.

Regarding the other comments: I tested with the ATs already. Mobile menu may be expanded by default, but that means there will be movement in the page when javascript kicks in and hides it.

nickcolley commented on src/components/header/template.njk:
```diff
@@ -0,0 +1,79 @@
+              fill="currentColor" fill-rule="evenodd"
+              d="M25 30.2c3.5 1.5 7.7-.2 9.1-3.7 1.5-3.6-.2-7.8-3.9-9.2-3.6-1.4-7.6.3-9.1 3.9-1.4 3.5.3 7.5 3.9 9zM9 39.5c3.6 1.5 7.8-.2 9.2-3.7 1.5-3.6-.2-7.8-3.9-9.1-3.6-1.5-7.6.2-9.1 3.8-1.4 3.5.3 7.5 3.8 9zM4.4 57.2c3.5 1.5 7.7-.2 9.1-3.8 1.5-3.6-.2-7.7-3.9-9.1-3.5-1.5-7.6.3-9.1 3.8-1.4 3.5.3 7.6 3.9 9.1zm38.3-21.4c3.5 1.5 7.7-.2 9.1-3.8 1.5-3.6-.2-7.7-3.9-9.1-3.6-1.5-7.6.3-9.1 3.8-1.3 3.6.4  ...
+            ></path>
+            {# Fallback PNG image for older browsers.
```
Excellent comments!

alex-ju replied:
@36degrees' work, not mine

hannalaakso commented on src/components/header/_header.scss:
```diff
@@ -0,0 +1,274 @@
+  .govuk-header__menu-button--open {
+    &::after {
+      @include govuk-shape-arrow($direction: up, $base: 10px);
+      display: inline-block;
```
It's a shame we have to set `display: inline-block` twice to override the mixin. Completely NOT a problem with the PR. I've raised an [issue.](https://github.com/alphagov/govuk-frontend/issues/698). It might worth linking to the issue in the code to say the second `display: inline-block` can be removed if the issue is fixed. Up to you @alex-ju 🙂

igloosi replied:
https://github.com/alphagov/govuk-frontend/pull/701

alex-ju replied:
@hannalaakso: @igloosi created a separate PR to deal with this

nickcolley commented on src/components/header/_header.scss:
```diff
@@ -0,0 +1,274 @@
+  }
+
+  @include mq($media-type: print) {
+    .app-header {
```
I think we may need the govuk- namespace

nickcolley commented on src/components/header/_header.scss:
```diff
@@ -0,0 +1,274 @@
+    display: block;
+  }
+
+  .govuk-header__navigation--right {
```
Could use `--end` instead of `--right` here

nickcolley commented on src/components/header/header.js:
```diff
@@ -0,0 +1,57 @@
+*/
+Header.prototype.toggleClass = function (node, className) {
+  if (node.classList.contains(className)) {
+    node.classList.remove(className)
```
I think we need a polyfill for IE9 support.

I don't think too many people would notice since IE9 users typically is used on desktop machines but better to be safe.

alex-ju commented:
@nickcolley: thanks for the review, good findings. updated addressing them.

nickcolley commented on src/components/header/template.njk:
```diff
@@ -0,0 +1,79 @@
+{% from "phase-banner/macro.njk" import govukPhaseBanner %}
```
I think we can remove this one

alex-ju commented:
@nickcolley Updated. Now the header relies on a `.js-enabled` class being added to the `<body>` (or actually any other parent element). Test for non-javascript behaviour added.

igloosi commented:
this looks very good now. can we ship it?

alex-ju commented:
@igloosi updated:
- Rebased to master
- Uses the new `govuk-shape-arrow` mixin
- Updated Changelog

P.S. I'm aware it needs some commit squashing

alex-ju commented:
Update: clean-up commit history
